RecoTauTag/RecoTau/plugins/DeepTauId.cc

```diff
@@ -414,6 +414,374 @@ namespace {
     }
   }  // namespace dnn_inputs_2017_v2
 
+  const std::map<std::string, size_t> matchDiscriminatorIndices(
```

Comment:
I agree that it would be good to avoid this. I am not sure I am following 100% what you mean with the enum solution however. Could you elaborate on this?

Reply:
True, I did not fully express my thought, also because I'm not sure that it is the best way to do it and I did not fully develop it. Thinking about it again, how about the following?: You can define an enum within the DeepTauID class listing the cutbased ID types. In the end of the matchDiscriminatorIndices function, you could call for each enum entry your map with the corresponding string (hardcoded, changing those would anyway require code changes in this class) and fill another map<enum,index> and return this one instead. This way you could already check, whether all needed inputs are available.
Alternatively and a bit shorter you could add some if-else to the matchDiscriminatorIndices to translate the strings from provenance into enum in order to fill the map. This would be a bit more direct, but also without check of completeness. But I'm also not sure how useful such a check would be.

Reply:
btw. with an enum you don't even need a map. You can use a vector of given size right away. This could be even more efficient. Just to be clear: The purpose of the enum would be to have human readible calls in your helper functions where you load the ID inputs.
If there was only one dR version, it might be even easier to define one index variable a class member for each ID type. But having this for each dR version would become a bit messy, I guess.
But you see that there are quite some options. Maybe you find something more beatiful.

Reply:
With 'check' I don't necessarily mean some super nice error message. The behaviour so far, where you would get an error that the map does not contain a certain string is actually fine already (I thought it's maybe nicer to get that during the initialization already, but in the event loop is actually fine as well). That's maybe the advantage of the map. If you switch to a vector, you'd need to take some extra care that all entries are filled properly in order to avoid undefined behavior.

Reply:
I hope I didn't increase confusion with my loose thoughts now.

Reply:
No, no problem, the input is definitely welcome. I was just thinking on how to properly do this. I am currently just struggling a bit to find a way to do this in a nice manner. Personally I think that the explicit string call in the maps in the helper functions make it fairly reasonable, but I do see your point.

Reply:
With the enum, your helper functions would look almost the same, just without the quotes around the names.

Reply:
that's why I suggested enum. For readability reasons, I would not like calls via plain intergers or sth like that, either.

Reply:
Ah yes, I just realized I was completely misinterpreting something in your explanation. I think we are on the same page now and I will make the changes